DistroLauncher/algorithms.h

```diff
@@ -18,40 +18,91 @@
 
 /// Common algorithms to be used everywhere in the launcher project with style resembling the std ones.
 
-/// Returns true if [tested] starts with [end]. Null-termination is not required.
-template <typename CharT>
-// NOLINTNEXTLINE(bugprone-easily-swappable-parameters) - Parameters must have the same type.
-bool starts_with(const std::basic_string_view<CharT> tested, const std::basic_string_view<CharT> start)
+namespace algo_internal
 {
-    if (tested.size() < start.size()) {
-        return false;
+    // Tells chars appart from other types
+    template <typename T> constexpr bool is_character() noexcept
+    {
+        return std::is_same_v<T, char> || std::is_same_v<T, wchar_t> || std::is_same_v<T, signed char> ||
+               std::is_same_v<T, unsigned char> || std::is_same_v<T, char16_t> || std::is_same_v<T, char32_t>;
+        // C++20 introduces char8_t
+    }
+
+    /**
+     * This template converts c-strings into string_views, and
+     * leaves all other types untouched
+     */
+    template <typename T> constexpr T&& string_adaptor(T&& value)
+    {
+        return std::forward<T>(value);
+    }
+
+    template <typename T> constexpr T const& string_adaptor(T const& value)
+    {
+        return value;
+    }
+
+    template <typename T> constexpr T& string_adaptor(T& value)
+    {
+        return value;
+    }
+
+    template <typename T> constexpr std::enable_if_t<!is_character<T>(), T*> string_adaptor(T* value)
+    {
+        return value;
+    }
+
+    template <typename T> constexpr std::enable_if_t<!is_character<T>(), T const*> string_adaptor(T const* value)
+    {
+        return value;
+    }
+
+    template <typename Char>
+    constexpr std::enable_if_t<is_character<Char>(), std::basic_string_view<Char>> string_adaptor(Char* value)
+    {
+        return {value};
+    }
+
+    template <typename Char>
+    constexpr std::enable_if_t<is_character<Char>(), std::basic_string_view<const Char>> string_adaptor(
+      Char const* value)
+    {
+        return {value};
+    }
+
+    template <typename IteratorTested, typename IteratorCompare>
+    bool starts_with_impl(IteratorTested tested_begin, IteratorTested tested_end, IteratorCompare compare_begin,
```

Comment:
clang-format must hate the standard library then 😓

Reply:
I don't disagree that those parameters are easily swappable and that would cause a silent bug. That's why named parameter support in C++ would be handy. And that's why C99/C++20 struct designated initializers are handy, because they allow mimicking named parameters. Even nicer it would be if the order didn't match. I hope C++ adopts such soon.